src/v/pandaproxy/schema_registry/json.cc

```diff
@@ -298,6 +300,454 @@ result<json::Document> parse_json(std::string_view v) {
+// for N is also valid for O. precondition: older and newer are both valid
+// schemas
+bool is_superset(json::Value const& older, json::Value const& newer);
+
+// close the implementation in a namespace to keep it contained
+namespace is_superset_impl {
+
+// helper struct to format json::Value
+struct pj {
+    json::Value const& v;
+    friend std::ostream& operator<<(std::ostream& os, pj const& p) {
+        auto osw = json::OStreamWrapper{os};
+        auto writer = json::Writer<json::OStreamWrapper>{osw};
+        p.v.Accept(writer);
+        return os;
+    }
+};
+
+enum class json_type : uint8_t {
+    string = 0,
+    integer = 1,
+    number = 2,
+    object = 3,
+    array = 4,
+    boolean = 5,
+    null = 6
+};
+// enough inlined space to hold all the values of json_type
+using json_type_list = absl::InlinedVector<json_type, 7>;
+
+constexpr auto json_schema_type_array
+  = std::to_array<std::pair<json_type, std::string_view>>(
+    {{json_type::string, "string"},
+     {json_type::integer, "integer"},
+     {json_type::number, "number"},
+     {json_type::object, "object"},
+     {json_type::array, "array"},
+     {json_type::boolean, "boolean"},
+     {json_type::null, "null"}});
+
+constexpr std::optional<json_type>
+schema_type_from_string(std::string_view sts) {
+    auto t_it = std::ranges::find(
+      json_schema_type_array, sts, [](auto t_s) { return t_s.second; });
+    if (t_it == json_schema_type_array.end()) {
+        return std::nullopt;
+    } else {
+        return t_it->first;
+    }
+}
+
+std::istream& operator>>(std::istream& is, json_type& t) {
+    auto sts = std::string{};
+    is >> sts;
+    auto opt_t = schema_type_from_string(sts);
+    if (opt_t.has_value()) {
+        t = opt_t.value();
+    } else {
+        is.setstate(std::ios::failbit);
+    }
+    return is;
+}
+
+// parse None | schema_type | array[schema_type] into a set of types.
+// the return type is implemented as a inlined_vector<json_type> with sorted set
+// semantics
+json_type_list normalized_type(json::Value const& v) {
+    auto type_it = v.FindMember("type");
+    auto ret = json_type_list{};
+    if (type_it == v.MemberEnd()) {
+        // omit keyword is like accepting all the types
+        std::ranges::copy(
+          json_schema_type_array | std::views::keys, std::back_inserter(ret));
+    } else if (type_it->value.IsArray()) {
+        // schema ensures that all the values are unique
+        std::ranges::copy(
+          type_it->value.GetArray()
+            | std::views::transform([](json::Value const& v) {
+                  return boost::lexical_cast<json_type>(
+                    std::string_view{v.GetString(), v.GetStringLength()});
+              }),
+          std::back_inserter(ret));
+    } else {
+        ret.push_back(boost::lexical_cast<json_type>(std::string_view{
+          type_it->value.GetString(), type_it->value.GetStringLength()}));
```

Comment:
nit: the parsing of a `json::Value` into a `json_type` is duplicated here

Reply:
what do you mean?

Reply:
The cast, I think, so something like my suggested `parse_json_type` would result in:
```c++
        std::ranges::copy(
          type_it->value.GetArray() | std::views::transform(&parse_json_type),
          std::back_inserter(ret));
    } else {
        ret.push_back(parse_json_type(type_it->value));
    }
```

Reply:
done